File: ios/bridge/ExpoBridgeLibrary.swift
import Foundation

@available(iOS 16.0, *)
enum ExpoBridgeLibrary {
  static func getUserPlaylists(service: LibraryService, options: NSDictionary) async throws -> [String: Any] {
    let pagination = BridgePagination(from: options)
    let playlists = try await service.getPlaylists(options: pagination)
    return BridgeResponses.playlists(playlists)
  }

  static func getLibrarySongs(service: LibraryService, options: NSDictionary) async throws -> [String: Any] {
    let pagination = BridgePagination(from: options)
    let songs = try await service.getSongs(options: pagination)
    return BridgeResponses.songs(songs)
  }

  static func getPlaylistSongs(service: LibraryService, playlistId: String) async throws -> [String: Any] {
    let songs = try await service.getPlaylistSongs(playlistId: playlistId)
    return BridgeResponses.songs(songs)
  }

  static func getLibraryArtists(service: LibraryService, options: NSDictionary) async throws -> [String: Any] {
    let pagination = BridgePagination(from: options)
    let artists = try await service.getArtists(options: pagination)
    return BridgeResponses.artists(artists)
  }

  static func getLibraryAlbums(service: LibraryService, options: NSDictionary) async throws -> [String: Any] {
    let pagination = BridgePagination(from: options)
    let albums = try await service.getAlbums(options: pagination)
    return BridgeResponses.albums(albums)
  }
}

File: android/src/main/java/expo/modules/applemusic/bridge/ExpoBridgeLibrary.kt
package expo.modules.applemusic.bridge

import expo.modules.applemusic.AndroidLibraryService
import expo.modules.applemusic.BridgeResponses
import expo.modules.applemusic.PaginationOptions
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.ModuleDefinitionBuilder

internal fun ModuleDefinitionBuilder.registerLibraryBridge(
  libraryService: () -> AndroidLibraryService,
) {
  AsyncFunction("getUserPlaylists") Coroutine { musicUserToken: String, options: Map<String, Any?> ->
    val pagination = PaginationOptions.fromMap(options)
    BridgeResponses.playlists(libraryService().getPlaylists(musicUserToken, pagination))
  }

  AsyncFunction("getLibrarySongs") Coroutine { musicUserToken: String, options: Map<String, Any?> ->
    val pagination = PaginationOptions.fromMap(options)
    BridgeResponses.songs(libraryService().getSongs(musicUserToken, pagination))
  }

  AsyncFunction("getPlaylistSongs") Coroutine {
      musicUserToken: String,
      playlistId: String,
      options: Map<String, Any?>,
    ->
    BridgeResponses.songs(libraryService().getPlaylistSongs(musicUserToken, playlistId))
  }

  AsyncFunction("getLibraryArtists") Coroutine { musicUserToken: String, options: Map<String, Any?> ->
    val pagination = PaginationOptions.fromMap(options)
    BridgeResponses.artists(libraryService().getArtists(musicUserToken, pagination))
  }

  AsyncFunction("getLibraryAlbums") Coroutine { musicUserToken: String, options: Map<String, Any?> ->
    val pagination = PaginationOptions.fromMap(options)
    BridgeResponses.albums(libraryService().getAlbums(musicUserToken, pagination))
  }

  AsyncFunction("getLibraryMusicVideos") Coroutine { musicUserToken: String, options: Map<String, Any?> ->
    val pagination = PaginationOptions.fromMap(options)
    BridgeResponses.musicVideos(libraryService().getMusicVideos(musicUserToken, pagination))
  }

  AsyncFunction("librarySearch") Coroutine {
      musicUserToken: String,
      term: String,
      types: List<String>,
      options: Map<String, Any?>,
    ->
    val pagination = PaginationOptions.fromMap(options)
    BridgeResponses.librarySearch(libraryService().search(musicUserToken, term, types, pagination))
  }
}
